bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -1066,5 +1059,73 @@ public String toString()
 
   }
 
+  private class ListenersNotifier implements Runnable
+  {
+    private boolean isMoreDataAvailable = false;
+
+    private synchronized void moreDataAvailable()
+    {
+      if (future == null) {
+        future = autoFlushExecutor.submit(this);
+      } else {
+        isMoreDataAvailable = true;
+      }
+    }
+
+    private boolean notifyListeners()
+    {
+      boolean doesAtLeastOneListenerHaveDataToSend = false;
+      for (DataListener dl : all_listeners) {
+        try {
+          doesAtLeastOneListenerHaveDataToSend |= dl.addedData(false);
+        } catch (RuntimeException e) {
+          logger.error("{}: removing DataListener {} due to exception", DataList.this, dl, e);
+          removeDataListener(dl);
+          break;
+        }
+      }
+      return doesAtLeastOneListenerHaveDataToSend;
+    }
+
+    private boolean checkIfListenersHaveDataToSendOnly()
+    {
+      synchronized (DataList.this) {
+        for (DataListener dl : all_listeners) {
+          try {
+            if (dl.addedData(true)) {
+              return true;
+            }
+          } catch (RuntimeException e) {
+            logger.error("{}: removing DataListener {} due to exception", DataList.this, dl, e);
+            removeDataListener(dl);
+            return checkIfListenersHaveDataToSendOnly();
+          }
+        }
+        return false;
+      }
+    }
+
+    @Override
+    public void run()
+    {
+      try {
+        if (notifyListeners() || checkIfListenersHaveDataToSendOnly()) {
```

Comment:
No, it requires two passes. The first pass is not synchronized and nodes/Subscribers are supposed to send as much data as they can, but it may take time. During the second path (synchronized) Subscribers are not supposed to send data or iterate over DataList.